src/Mafia/Cabal.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Mafia.Cabal
+  ( SandboxDir
+  , CabalError(..)
+  , repairIndexFile
+  ) where
+
+import           Control.Exception (IOException)
+import           Control.Monad.Catch (catch)
+import           Control.Monad.IO.Class (MonadIO(..))
+
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString.Lazy as L
+import qualified Data.Text.Encoding as T
+
+import qualified Codec.Archive.Tar as Tar
+
+import           Mafia.IO
+import           Mafia.Path
+
+import           P
+
+import           System.IO (IO)
+
+import           X.Control.Monad.Trans.Either
+
+------------------------------------------------------------------------
+
+type SandboxDir = Directory
+
+data CabalError =
+    IndexFileNotFound File
+  | CorruptIndexFile Tar.FormatError
+  deriving (Show)
+
+------------------------------------------------------------------------
+
+readIndexFile :: MonadIO m => File -> EitherT CabalError m ByteString
+readIndexFile file = do
+  mbs <- readBytes file
+  hoistEither $ maybe (Left (IndexFileNotFound file)) Right mbs
+
+readEntries :: Monad m => ByteString -> EitherT CabalError m [Tar.Entry]
+readEntries bs = do
+  let tar = Tar.read (L.fromStrict bs)
+  hoistEither $ Tar.foldEntries (\x -> fmap (x :)) (Right []) (Left . CorruptIndexFile) tar
+
+isTreeRef :: Tar.Entry -> Bool
+isTreeRef x = Tar.entryPath x == "local-build-tree-reference/"
+
+fromEntry :: Tar.Entry -> Maybe Directory
+fromEntry x =
+  case Tar.entryContent x of
+    Tar.OtherEntryType _ lbs _ | isTreeRef x -> Just (T.decodeUtf32LE (L.toStrict lbs))
+    _                                        -> Nothing
+
+------------------------------------------------------------------------
+
+repairIndexFile :: SandboxDir -> EitherT CabalError IO ()
+repairIndexFile = filterEntries doesDirectoryExist
+
+filterEntries :: (Directory -> IO Bool) -> File -> EitherT CabalError IO ()
+filterEntries pred sandboxDir = do
+  let indexFile  = sandboxDir </> "packages/00-index.tar"
```

Comment:
As discussed not 100% whether it makes sense to fail hard if this file doesn't exist or not. On the one hand it means we can spot issues quicker, but I'm vaguely thinking of cabal's track record of breaking compatibility and what that would mean for `mafia` projects one year from now. I'm hoping will be a different beast by then anyway, so just ignore me. :)